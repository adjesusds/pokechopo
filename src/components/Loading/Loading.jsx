// External Packages
import { number, string } from 'prop-types'
import React, { Component } from 'react'

/**
 * Creates a new Loading component.
 * @class
 */
class Loading extends Component {
  /** @override */
  static propTypes = {
    message: string,
    suffix: string,
    suffixDuration: number,
    suffixLength: number,
  }

  /** @override */
  static defaultProps = {
    message: 'Loading',
    suffix: '.',
    suffixDuration: 1000,
    suffixLength: 3,
  }

  /** @override */
  state = {
    ticks: 0,
  }

  /** @override */
  componentDidMount() {
    this.tickInterval = setInterval(this.tick, this.props.suffixDuration)
  }

  /** @override */
  componentWillUnmount() {
    clearInterval(this.tickInterval)
  }

  /** @private */
  tick = () => {
    this.setState(prevState => ({
      ticks: prevState.ticks < this.props.suffixLength ? prevState.ticks + 1 : 0,
    }))
  }

  /** @override */
  render() {
    return (
      <span>
        {this.props.message}
        {this.props.suffix.repeat(this.state.ticks)}
      </span>
    )
  }
}

export default Loading
