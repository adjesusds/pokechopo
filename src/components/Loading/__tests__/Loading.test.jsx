// External Packages
import { expect as chaiExpect } from 'chai'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { spy } from 'sinon'

// Internal Packages
import Loading from '../Loading'

describe('<Loading />', () => {
  it('should render the component on valid props', () => {
    const component = shallow(<Loading />)

    expect(component).toMatchSnapshot()
  })

  it('should update the component on valid actions', () => {
    const component = mount(<Loading />)

    chaiExpect(component.state().ticks).to.equal(0)

    component.instance().tick()

    chaiExpect(component.state().ticks).to.equal(1)

    component.instance().tick()

    chaiExpect(component.state().ticks).to.equal(2)

    component.instance().tick()

    chaiExpect(component.state().ticks).to.equal(3)

    component.instance().tick()

    chaiExpect(component.state().ticks).to.equal(0)
  })

  it('should unmount the component on valid actions', () => {
    spy(Loading.prototype, 'componentWillUnmount')

    const component = mount(<Loading />)

    component.unmount()

    chaiExpect(Loading.prototype.componentWillUnmount).to.have.property('callCount', 1)

    Loading.prototype.componentWillUnmount.restore()
  })
})
