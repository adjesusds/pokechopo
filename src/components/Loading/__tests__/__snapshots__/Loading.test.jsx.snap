// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> should render the component on valid props 1`] = `
<span>
  Loading
</span>
`;
